builtin/providers/aws/resource_aws_s3_bucket.go

```diff
@@ -244,6 +251,9 @@ func resourceAwsS3BucketRead(d *schema.ResourceData, meta interface{}) error {
 	if err := d.Set("website_endpoint", endpoint); err != nil {
 		return err
 	}
+	if err := d.Set("website_domain", strings.TrimPrefix(endpoint, fmt.Sprintf("%s.", d.Id()))); err != nil {
```

Comment:
Instead of trimming a string we build elsewhere, why not build the domain first and use that domain in `website_endpoint`?
Considering we have to call the `s3conn.GetBucketLocation` (i.e. we need the bucket name to get the domain), I would consider creating a new `type`, e.g.

```
type S3Website struct {
  Endpoint, Domain string
}
```

and then change `websiteEndpoint(s3conn *s3.S3, d *schema.ResourceData) (string, error)` to `website(s3conn *s3.S3, d *schema.ResourceData) (S3Website, error)`.

That should be efficient enough since we only need to call the function once and get both domain+endpoint.